Fix Conan Source folders.root issue

timviola-cmi opened the pull request:
Changelog: Bugfix: Fixes issue where conanfile's `source()` method doesn't use `folders.root` when present.
Docs: Omit

Close #18376


No doc update should be needed as this implements the expected behavior for folders.root.

timviola-cmi committed:
Fix Conan Source folders.root issue

- Fixed issue in local.py where source() method did not account for
  self.folders.root (when specified)

- Added
test/integration/command/source_test.py::SourceTest::test_local_source_layout_root

Close #18376

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=18377) <br/>All committers have signed the CLA.

memsharded commented:
Hi @timviola-cmi 

Thanks for your contribution.
It would be necessary a couple of things to be able to consider this:
- First, adding a red/green test that covers this change. Please let us know if you need guidance with this, we can help.
- Signing the CLA, could you please check the CLA above?

Many thanks!

timviola-cmi commented:
Added a test that verifies the source method applies the root when called locally

memsharded commented on test/integration/command/source_test.py:
```diff
@@ -142,6 +142,30 @@ def source(self):
+            class ConanLib(ConanFile):
+                
+                def layout(self):
+                    self.folders.root = "root_folder"
```
From docs in: https://docs.conan.io/2/reference/conanfile/methods/layout.html

> self.folders.root (Defaulted to None): Specifies a parent directory where the sources, generators, etc., are located specifically when the conanfile.py is located in a separated subdirectory. Check [this example](https://docs.conan.io/2/examples/conanfile/layout/conanfile_in_subfolder.html#examples-conanfile-layout-conanfile-in-subfolder) on how to use self.folders.root.

So it seems this layout is not allowed. The ``self.folders.root`` is intended for solving other issues, like the possibility of having the ``conanfile.py`` inside a subfolder instead of having it in the root of the repo.

What problem are you trying to solve with this layout?
Sorry I didn't realized it in the ticket https://github.com/conan-io/conan/issues/18376, I'll put a note there, and maybe conversation can follow there too.

memsharded commented on test/integration/command/source_test.py:
```diff
@@ -142,6 +142,30 @@ def source(self):
+        client.save({CONANFILE: conanfile})
+
+        client.run("source .")
+        self.assertTrue(client.out.strip().endswith("spaces\\root_folder"))
```
This is most likely broken in other systems such as Linux, it seems you developed and tested it under Windows?
In any case, lets go back to the issue, to clarify the goal and the intent, and then we will go back to the PR code, thanks!

timviola-cmi replied:
Sounds good. Thanks for taking the time to review. I did test under windows. 

## The example from the documentation fails 
When you run the example you can see that the `source()` method fails because it is looking for *CMakeLists.txt* in the `conan` directory. You can prove this by moving *CMakeLists.txt* into the `conan` directory and then `conan source .` will run successfully.
![image](https://github.com/user-attachments/assets/33f06a75-40a5-42ff-8f21-953b3b87d688)

## Conditions
I did see in the documentation that it said parent folder, but this fails even if following the parent folder guidance as mentioned above. The particular conditions under which this really matters:
1. The conanfile is not stored in the same repository as the source
1. You cannot use `export_sources()` because the source code should not be available to all package consumers
1. `no_copy_source=True` is used because the size of the repository is significant and the copy time is non_trivial (thus even in the `build()` method `self.source_folder` must be used)

## Failure when using parent folder

For example assuming a folder structure as follows:
![image](https://github.com/user-attachments/assets/8713c729-1dfe-485a-a430-cc608289da0d)

Consider the following conanfile (which conforms to the parent folder guidance):
```python
from conan import ConanFile
from conan.tools.files import save, load

class BasicConanfile(ConanFile):
    name = "conan_source_bug"
    version = "0.1"
    description = "Demonstrate a bug in conan source() method"
    no_copy_source = True
    
    def layout(self):
        self.folders.root = ".."
        self.folders.source = "src"
        self.folders.generators = "generated"
    
    def generate(self):
        save(self, "generated_file.txt", "content")
        
    def source(self):
        # Saves source to source_folder
        self.output.info(f"In the source() method the Source folder is: {self.source_folder}")
        save(self, "source_file.c", "content") #i.e. get source from somewhere

    def build(self):
        self.output.info(f"In the build() method the Source folder is: {self.source_folder}")
        load(self, f"{self.source_folder}/source_file.c") # i.e. a compiler looking for source files
```

If you attempt to run conan source -> conan install -> conan build it will fail because the build method has a different definition of source_folder.
![image](https://github.com/user-attachments/assets/529f8447-6aaa-44c4-a746-6d1b7f78cb00)
![image](https://github.com/user-attachments/assets/380638dd-8639-46b1-bc45-5ecfc34549d3)

Correct me if I'm misunderstanding something.

## Use case for having a root that is a subdir instead of parent

We ran across a use case where it t was desired that the `source()` method and all other methods use a root that is a subfolder of the current directory. This is our preferred workspace setup:
- ParentDir
    - conanfile.py
    - root_folder
        - src
        - generated  

From a unix "principle of least surprise" perspective, it seems root folder should function consistently whether it is a subdirectory or parent folder and I don't think it requires any additional code to support both cases.

The fix I submitted is agnostic to whether it's a parent or child folder (the above conanfile will also work with the submitted fix). 

Let me know if I've missed anything!

memsharded commented on test/integration/command/source_test.py:
```diff
@@ -142,6 +142,30 @@ def source(self):
+        client.save({CONANFILE: conanfile})
+
+        client.run("source .")
+        self.assertTrue(client.out.strip().endswith("spaces\\root_folder"))
```
I think it would be better to make the recipe read a file in the source() and build() method in the test, instead of printing the source folder. If those methods fail to find the file, then it will fail raising an error, I think that test will be easier than having to check the source folder path.

memsharded commented on test/integration/command/source_test.py:
```diff
@@ -142,6 +142,30 @@ def source(self):
         self.assertIn("conanfile.py: Running source!", client.out)
         self.assertEqual("Hello World", client.load("file1.txt"))
 
+    def test_local_source_layout_root(self):
```
I think I'd change this test to implement the recipe that you provided, with the ``self.folders.root = ".."``. That helps to tag this as a ``Bugfix`` and be able to move it forward.

Then, I'd move new tests to cover the use case of a child "root" folder to a different PR, to be done later, once this is merged.

timviola-cmi replied:
Ok, will adjust the tests tomorrow.

memsharded reviewed (approved):
Please ask for guidance or help with the test if necessary. Thanks again for your contribution!

Tagging this for next 2.18 release

timviola-cmi committed:
Update Integration Test

- Test should now pass on windows, linux, and mac OS
- Test creates a file in `source()` method and loads the file
  in the `build()` method. The test will fail if both `source()`
  and `build()` do not use the same `self.source_folder`

memsharded reviewed (approved):
Looking good, many thanks for your contribution!
Merging this, it will be in next 2.18 release

memsharded commented on test/integration/command/source_test.py:
```diff
@@ -142,6 +142,39 @@ def source(self):
+        # The build method will fail if source() and build() don't 
+        # both use the correct source_folder
+        client.run("build .")
+        print(client.out)
```
```suggestion
        assert "conanfile.py: Calling build()" in client.out  # doesn't fail
```

memsharded committed:
Update test/integration/command/source_test.py

memsharded committed:
Merge branch 'develop2' into fix/conansourcerootfolder
